File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/model/factory/BaseTestStepFactory.java
package org.skellig.teststep.processing.model.factory;

import org.skellig.teststep.processing.converter.TestDataConverter;
import org.skellig.teststep.processing.converter.TestStepValueConverter;
import org.skellig.teststep.processing.model.TestStep;
import org.skellig.teststep.processing.model.ValidationDetails;
import org.skellig.teststep.processing.utils.CachedPattern;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class BaseTestStepFactory implements TestStepFactory {


    private static final String TEST_STEP_NAME_KEYWORD = "test.step.keyword.name";
    private static final String VARIABLES_KEYWORD = "test.step.keyword.variables";

    private static Set<String> testDataKeywords;

    private Properties keywordsProperties;
    private TestDataConverter testDataConverter;
    private TestStepFactoryValueConverter testStepFactoryValueConverter;
    private ValidationDetailsFactory validationDetailsFactory;

    public BaseTestStepFactory(Properties keywordsProperties,
                               TestStepValueConverter testStepValueConverter,
                               TestDataConverter testDataConverter) {

        this.keywordsProperties = keywordsProperties;
        this.testStepFactoryValueConverter = new TestStepFactoryValueConverter(testStepValueConverter);
        this.validationDetailsFactory = new ValidationDetailsFactory(keywordsProperties, testStepFactoryValueConverter);
        this.testDataConverter = testDataConverter;

        if (testDataKeywords == null) {
            testDataKeywords = Stream.of(
                    getKeywordName("test.step.keyword.data", "data"),
                    getKeywordName("test.step.keyword.payload", "payload"),
                    getKeywordName("test.step.keyword.body", "body"),
                    getKeywordName("test.step.keyword.request", "request"),
                    getKeywordName("test.step.keyword.response", "response"),
                    getKeywordName("test.step.keyword.message", "message"))
                    .collect(Collectors.toSet());
        }

    }

    @Override
    public TestStep create(String testStepName, Map<String, Object> rawTestStep, Map<String, String> parameters) {
        Map<String, Object> additionalParameters = new HashMap<>(parameters);

        Map<String, String> parametersFromTestName = extractParametersFromTestStepName(testStepName, rawTestStep);
        if (parametersFromTestName != null) {
            additionalParameters.putAll(parametersFromTestName);
        }

        Map<String, Object> variables = extractVariables(rawTestStep, additionalParameters);
        if (variables != null) {
            additionalParameters.putAll(variables);
        }

        return createTestStep(rawTestStep).create(
                getId(rawTestStep),
                testStepName,
                extractTestData(rawTestStep, additionalParameters),
                validationDetailsFactory.create(rawTestStep, additionalParameters),
                additionalParameters,
                variables);
    }

    protected abstract CreateTestStepDelegate createTestStep(Map<String, Object> rawTestStep);

    private Map<String, Object> extractVariables(Map<String, Object> rawTestStep, Map<String, Object> parameters) {
        Object rawVariables = rawTestStep.get(getKeywordName(VARIABLES_KEYWORD, "variables"));
        Object convertedVariables = convertHierarchicalData(rawVariables, parameters);
        return convertedVariables instanceof Map ? (Map<String, Object>) convertedVariables : null;
    }

    private Object extractTestData(Map<String, Object> rawTestStep, Map<String, Object> parameters) {
        return testDataKeywords.stream()
                .filter(rawTestStep::containsKey)
                .map(keyword -> testDataConverter.convert(convertHierarchicalData(rawTestStep.get(keyword), parameters)))
                .findFirst()
                .orElse(null);
    }

    private Object convertHierarchicalData(Object data, Map<String, Object> parameters) {
        if (data instanceof Map) {
            return ((Map<String, Object>) data).entrySet().stream()
                    .collect(Collectors.toMap(Map.Entry::getKey, entry -> convertHierarchicalData(entry.getValue(), parameters)));
        } else if (data instanceof List) {
            return ((List) data).stream()
                    .map(item -> convertHierarchicalData(item, parameters))
                    .collect(Collectors.toList());
        } else {
            return testStepFactoryValueConverter.convertValue(data, parameters);
        }
    }

    private Map<String, String> extractParametersFromTestStepName(String testStepName, Map<String, Object> rawTestStep) {
        Map<String, String> parameters = null;
        Matcher matcher = CachedPattern.compile(getName(rawTestStep)).matcher(testStepName);
        if (matcher.find()) {
            parameters = new HashMap<>();
            for (int i = 1; i <= matcher.groupCount(); i++) {
                parameters.put("$" + i, matcher.group(i));
            }
        }
        return parameters;
    }

    protected String getId(Map<String, Object> rawTestStep) {
        return rawTestStep.containsKey("id") ? String.valueOf(rawTestStep.get("id")) : null;
    }

    protected String getName(Map<String, Object> rawTestStep) {
        return String.valueOf(rawTestStep.get(getKeywordName(TEST_STEP_NAME_KEYWORD, "name")));
    }

    protected String getKeywordName(String keywordName, String defaultValue) {
        return keywordsProperties == null ? defaultValue : keywordsProperties.getProperty(keywordName, defaultValue);
    }

    protected <T> T convertValue(Object value, Map<String, Object> parameters) {
        return testStepFactoryValueConverter.convertValue(value, parameters);
    }

    protected interface CreateTestStepDelegate {

        TestStep create(String id, String name, Object testData, ValidationDetails validationDetails,
                        Map<String, Object> parameters, Map<String, Object> variables);
    }

}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/model/factory/BaseTestStepFactory.kt
package org.skellig.teststep.processing.model.factory

import org.skellig.teststep.processing.model.TestStep
import org.skellig.teststep.processing.util.CachedPattern
import org.skellig.teststep.processing.value.ValueExpressionContextFactory
import org.skellig.teststep.reader.value.expression.AlphanumericValueExpression
import org.skellig.teststep.reader.value.expression.ValueExpression

/**
 * BaseTestStepFactory is an abstract class that implements the TestStepFactory interface.
 * It provides common functionality and helper methods for creating TestStep objects.
 *
 * @param T The type of TestStep.
 * @property valueExpressionContextFactory A factory to create a ValueExpressionContext for evaluating [ValueExpression].
 */
abstract class BaseTestStepFactory<T : TestStep>(val valueExpressionContextFactory: ValueExpressionContextFactory) : TestStepFactory<T> {

    companion object {
        @JvmStatic
        protected fun fromProperty(property: String) : AlphanumericValueExpression = AlphanumericValueExpression(property)
        protected val TEST_STEP_NAME_KEYWORD = fromProperty("name")
    }

    protected open fun <T> convertValue(value: ValueExpression?, parameters: Map<String, Any?>): T? {
        return value?.evaluate(valueExpressionContextFactory.create(parameters)) as T?
    }

    /**
     * Extracts parameters from the test step name by matching it against a regular expression pattern.
     * If the regex pattern of a test step has groups, and you apply already formed name of the test step, then
     * it extract values from these groups and add them to the existing Map of parameters, where key is a respective group number,
     * and value is extracted value from the regex group. For example:
     *
     * If you have a test step in a file with this name: "User (.+) sends (\d+) coins"
     * and you run a test: "User Alex sends 100 coins", then you'll get these additional parameters:
     * - "1" - Alex
     * - "2"- 100
     *
     * which you can reference in the test step "User (.+) sends (\d+) coins" using expressions such as: ${1} or ${2}
     *
     * @param testStepName The name of the test step.
     * @param rawTestStep The raw test step, which is a map of value expressions and their corresponding values (nullable).
     * @return A mutable map of extracted parameters, where the key is the group number and the value is the extracted value.
     *         Returns null if no parameters were extracted. The Map is mutable because it may be modified within a test step.
     *         (see. TaskTestStepProcessor)
     */
    protected fun extractParametersFromTestStepName(testStepName: String, rawTestStep: Map<ValueExpression, ValueExpression?>): MutableMap<String, Any?>? {
        var parameters: MutableMap<String, Any?>? = null
        val matcher = CachedPattern.compile(getName(rawTestStep)).matcher(testStepName)
        if (matcher.find()) {
            parameters = mutableMapOf()
            for (i in 1..matcher.groupCount()) {
                val value = matcher.group(i)
                if (value.isNotEmpty()) {
                    parameters[i.toString()] = value
                }
            }
        }
        return parameters
    }

    /**
     * Returns the name from the raw test step.
     *
     * @param rawTestStep The raw test step.
     * @return The name of the test step as a string.
     */
    protected open fun getName(rawTestStep: Map<ValueExpression, ValueExpression?>): String {
        // No need to evaluate as it is always the value from toString
        return rawTestStep[TEST_STEP_NAME_KEYWORD].toString()
    }


}